File: togo/internal/cli/remove.go
package cli

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/nikeee/cappu/internal/config"
)

// RunRemove handles `cappu remove <configuration> <group:artifact>...`: drop the
// entries from cappu.json (comments preserved), then re-resolve and rewrite the
// lock + .cappu/lib like `cappu add` in reverse. The version segment of a
// coordinate is ignored - a dependency is removed by its group:artifact key.
// Port of src/cli/remove.ts.
func RunRemove(configurationArg string, specs []string, configPathArg string, cfg *config.Config) int {
	configuration := resolveConfiguration(configurationArg)
	keys := make([]string, 0, len(specs))
	var invalid []string
	for _, spec := range specs {
		if c, ok := parseAddCoordinate(spec); ok {
			keys = append(keys, c.key)
		} else {
			invalid = append(invalid, spec)
		}
	}
	if configuration == "" || len(keys) == 0 || len(invalid) > 0 {
		for _, spec := range invalid {
			fmt.Fprintf(os.Stderr, "cappu: not a coordinate: '%s'\n", spec)
			emitAnnotation("error", fmt.Sprintf("not a coordinate: '%s'", spec), AnnotationLocation{})
		}
		fmt.Fprintf(os.Stderr, "usage: cappu remove <%s> <group:artifact> [more...]\n"+
			"       aliases: a=api, i=implementation, ap=annotationProcessor, ti=testImplementation\n",
			strings.Join(configurations, "|"))
		return 2
	}
	if !cfg.FromFile {
		fmt.Fprintln(os.Stderr, "cappu: no cappu.json found - run `cappu init` first")
		emitAnnotation("error", "no cappu.json found - run `cappu init` first", AnnotationLocation{})
		return 1
	}

	configPath := filepath.Join(cfg.BaseDir, config.DefaultConfigName)
	if configPathArg != "" {
		if abs, err := filepath.Abs(configPathArg); err == nil {
			configPath = abs
		}
	}

	text, err := os.ReadFile(configPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "cappu: %s\n", err)
		return 1
	}
	any := false
	for _, key := range keys {
		next, removed, err := config.RemoveDependency(text, configuration, key)
		if err != nil {
			fmt.Fprintf(os.Stderr, "cappu: %s\n", err)
			return 1
		}
		if removed {
			text = next
			any = true
			fmt.Fprintf(os.Stderr, "removed %s %s\n", configuration, key)
		} else {
			fmt.Fprintf(os.Stderr, "warning: %s is not a %s dependency\n", key, configuration)
			emitAnnotation("warning", fmt.Sprintf("%s is not a %s dependency", key, configuration), AnnotationLocation{})
		}
	}
	if !any {
		return 1
	}
	if err := os.WriteFile(configPath, text, 0o644); err != nil {
		fmt.Fprintf(os.Stderr, "cappu: %s\n", err)
		return 1
	}

	reloaded, err := config.Load(configPath, "")
	if err != nil {
		fmt.Fprintf(os.Stderr, "cappu: %s\n", err)
		return 1
	}
	return runInstallWith(reloaded, false, true)
}

File: src/cli/remove.ts
// `cappu remove <configuration> <group:artifact>`: drop the entry from the
// cappu.json dependencies section (preserving comments - the file is JSONC) and
// then re-resolve and rewrite the lock + `.cappu/lib`, exactly like `cappu add`
// in reverse. The version segment of a `group:artifact[:version]` coordinate is
// ignored - a dependency is removed by its `group:artifact` key.

import { readFileSync, writeFileSync } from "node:fs";
import { join, resolve } from "node:path";

import {
  type CappuConfig,
  DEFAULT_CONFIG_NAME,
  DEPENDENCY_CONFIGURATIONS,
  loadConfig,
} from "../config.ts";
import { emitAnnotation } from "./annotations.ts";
import { parseAddCoordinate, resolveConfiguration } from "./add.ts";
import { runInstall } from "./install.ts";
import { removeJsoncKey } from "./jsoncEdit.ts";

const CONFIGURATIONS = DEPENDENCY_CONFIGURATIONS;
type Configuration = (typeof CONFIGURATIONS)[number];

/**
 * Delete the dependency from the JSONC config text (comments intact). Returns
 * the new text and whether the key was actually present.
 */
export function removeDependencyFromJsonc(
  text: string,
  configuration: Configuration,
  key: string,
): { text: string; removed: boolean } {
  return removeJsoncKey(text, ["dependencies", configuration, key]);
}

export async function runRemove(
  configurationArg: string | undefined,
  specs: readonly string[],
  configPathArg: string | undefined,
  config: CappuConfig,
): Promise<never> {
  const configuration = resolveConfiguration(configurationArg);
  // Accept `group:artifact` or `group:artifact:version` (the version is ignored).
  const keys = specs.map(s => parseAddCoordinate(s)?.key);
  const invalid = specs.filter((_, i) => keys[i] === undefined);
  if (!configuration || keys.length === 0 || invalid.length > 0) {
    for (const spec of invalid) {
      process.stderr.write(`cappu: not a coordinate: '${spec}'\n`);
      emitAnnotation("error", `not a coordinate: '${spec}'`);
    }
    process.stderr.write(
      `usage: cappu remove <${CONFIGURATIONS.join("|")}> <group:artifact> [more...]\n` +
        "       aliases: a=api, i=implementation, ap=annotationProcessor, ti=testImplementation\n",
    );
    process.exit(2);
  }
  if (!config.fromFile) {
    process.stderr.write("cappu: no cappu.json found - run `cappu init` first\n");
    emitAnnotation("error", "no cappu.json found - run `cappu init` first");
    process.exit(1);
  }

  const configPath = configPathArg
    ? resolve(configPathArg)
    : join(config.baseDir, DEFAULT_CONFIG_NAME);

  let text = readFileSync(configPath, "utf8");
  let any = false;
  for (const key of keys as string[]) {
    const result = removeDependencyFromJsonc(text, configuration, key);
    if (result.removed) {
      text = result.text;
      any = true;
      process.stderr.write(`removed ${configuration} ${key}\n`);
    } else {
      process.stderr.write(`warning: ${key} is not a ${configuration} dependency\n`);
      emitAnnotation("warning", `${key} is not a ${configuration} dependency`);
    }
  }

  if (!any) process.exit(1);
  writeFileSync(configPath, text);

  // Re-resolve and rewrite the lock against the reduced dependency set.
  return runInstall(loadConfig(configPath), { updateLock: true });
}
